File: src/demo_tf/scripts/publish_relation.py
#! /usr/bin/env python3

import rospy
import sys
from turtlesim.msg import Pose
from geometry_msgs.msg import TransformStamped
import tf2_ros
import tf_conversions

turtle_name = ""

def doPose(pose):

    broadcaster = tf2_ros.TransformBroadcaster()

    tfs = TransformStamped()
    tfs.header.frame_id = "world"
    tfs.header.stamp = rospy.Time.now()

    tfs.child_frame_id = turtle_name
    tfs.transform.translation.x = pose.x
    tfs.transform.translation.y = pose.y
    tfs.transform.translation.z = 0.0

    qtn = tf_conversions.transformations.quaternion_from_euler(0, 0, pose.theta)
    tfs.transform.rotation.x = qtn[0]
    tfs.transform.rotation.y = qtn[1]
    tfs.transform.rotation.z = qtn[2]
    tfs.transform.rotation.w = qtn[3]

    broadcaster.sendTransform(tfs)


if __name__ == "__main__":

    rospy.init_node("sub_tfs_p")

    rospy.loginfo("-------------------------------%d", len(sys.argv))
    if len(sys.argv) < 2:
        rospy.loginfo("Please pass in the parameter: the turtle namespace")
    else:
        turtle_name = sys.argv[1]
    rospy.loginfo("Turtle is: %s", turtle_name)

    rospy.Subscriber(turtle_name + "/pose",Pose,doPose)

    rospy.spin()

File: src/demo_tf/src/publish_relation.cpp
#include "ros/ros.h"
#include "turtlesim/Pose.h"
#include "tf2_ros/transform_broadcaster.h"
#include "tf2/LinearMath/Quaternion.h"
#include "geometry_msgs/TransformStamped.h"

std::string turtle_name;

void doPose(const turtlesim::Pose::ConstPtr& pose)
{
    static tf2_ros::TransformBroadcaster broadcaster;

    geometry_msgs::TransformStamped tfs;
    tfs.header.frame_id = "world";
    tfs.header.stamp = ros::Time::now();
    tfs.child_frame_id = turtle_name;
    tfs.transform.translation.x = pose->x;
    tfs.transform.translation.y = pose->y;
    tfs.transform.translation.z = 0.0;
    tf2::Quaternion qtn;
    qtn.setRPY(0,0,pose->theta);
    tfs.transform.rotation.x = qtn.getX();
    tfs.transform.rotation.y = qtn.getY();
    tfs.transform.rotation.z = qtn.getZ();
    tfs.transform.rotation.w = qtn.getW();

    broadcaster.sendTransform(tfs);
} 

int main(int argc, char *argv[])
{
    ros::init(argc,argv,"pub_tf");

    if (argc != 2)
    {
        ROS_ERROR("Please pass in the correct parameters");
    }
    else
    {
        turtle_name = argv[1];
        ROS_INFO("Turtle %s launched", turtle_name.c_str());
    }

    ros::NodeHandle nh;

    ros::Subscriber sub = nh.subscribe<turtlesim::Pose>(turtle_name + "/pose", 1000, doPose);

    ros::spin();

    return 0;
}